crates/sui-bridge-indexer/src/eth_worker.rs

```diff
@@ -1,21 +1,162 @@
+use crate::postgres_manager::get_latest_eth_token_transfer;
 use crate::postgres_manager::{write, PgPool};
 use crate::{BridgeDataSource, TokenTransfer, TokenTransferData, TokenTransferStatus};
+use anyhow::Result;
 use ethers::providers::Provider;
 use ethers::providers::{Http, Middleware};
 use ethers::types::Address as EthAddress;
+use mysten_metrics::spawn_logged_monitored_task;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::str::FromStr;
 use std::sync::Arc;
 use sui_bridge::abi::{EthBridgeEvent, EthSuiBridgeEvents};
 use sui_bridge::types::EthLog;
+use sui_bridge::{eth_client::EthClient, eth_syncer::EthSyncer};
+use tokio::task::JoinHandle;
 use tracing::info;
 use tracing::log::error;
 
-pub async fn process_eth_events(
-    mut eth_events_rx: mysten_metrics::metered_channel::Receiver<(EthAddress, u64, Vec<EthLog>)>,
+#[derive(Clone, Debug)]
+pub struct EthBridgeWorker {
+    provider: Arc<Provider<Http>>,
+    pg_pool: PgPool,
+    metrics: BridgeIndexerMetrics,
+    bridge_address: EthAddress,
+    config: Config,
+}
+
+impl EthBridgeWorker {
+    pub fn new(pg_pool: PgPool, metrics: BridgeIndexerMetrics, config: Config) -> Self {
+        let bridge_address = match EthAddress::from_str(&config.eth_sui_bridge_contract_address) {
+            Ok(addr) => addr,
+            Err(e) => {
+                panic!("Invalid Ethereum address: {}", e);
+            }
+        };
+
+        let provider = Arc::new(
+            Provider::<Http>::try_from(&config.eth_rpc_url)
+                .unwrap_or_else(|_| {
+                    panic!(
+                        "Cannot create Ethereum HTTP provider, URL: {}",
+                        &config.eth_rpc_url
+                    )
+                })
+                .interval(std::time::Duration::from_millis(2000)),
+        );
+        Self {
+            provider,
+            pg_pool,
+            metrics,
+            bridge_address,
+            config,
+        }
+    }
+
+    pub async fn start_indexing_finalized_events(&self) -> Result<JoinHandle<()>> {
+        let eth_client = Arc::new(
+            EthClient::<ethers::providers::Http>::new(
+                &self.config.eth_rpc_url,
+                HashSet::from_iter(vec![self.bridge_address]),
+            )
+            .await?,
+        );
+
+        let newest_finalized_block = match get_latest_eth_token_transfer(&self.pg_pool, true)? {
+            Some(transfer) => transfer.block_height as u64,
+            None => self.config.start_block,
+        };
+
+        info!("Starting from finalized block: {}", newest_finalized_block);
+
+        let finalized_contract_addresses =
+            HashMap::from_iter(vec![(self.bridge_address, newest_finalized_block)]);
+
+        let (_task_handles, eth_events_rx, _) =
+            EthSyncer::new(eth_client, finalized_contract_addresses)
+                .run()
+                .await
+                .expect("Failed to start eth syncer");
```

Comment:
let's fix this while we are working on it. Return the error instead of panicking here